app/antedecorators/gas.go

```diff
@@ -32,22 +35,112 @@ func GetGasMeterSetter(aclkeeper aclkeeper.Keeper) func(bool, sdk.Context, uint6
 }
 
 func getMessageMultiplierDenominator(ctx sdk.Context, msg sdk.Msg, aclKeeper aclkeeper.Keeper) uint64 {
+	// TODO: reason through whether it's reasonable to require non-* identifier for all operations
+	//       under the context of inter-contract changes
+	// only give gas discount if none of the dependency (except COMMIT) has id "*"
 	if wasmExecuteMsg, ok := msg.(*wasmtypes.MsgExecuteContract); ok {
-		addr, err := sdk.AccAddressFromBech32(wasmExecuteMsg.Contract)
-		if err != nil {
-			return DefaultGasMultiplierDenominator
-		}
-		mapping, err := aclKeeper.GetWasmDependencyMapping(ctx, addr, "", []byte{}, false, make(aclkeeper.ContractReferenceLookupMap))
-		if err != nil {
-			return DefaultGasMultiplierDenominator
-		}
-		// only give gas discount if none of the dependency (except COMMIT) has id "*"
-		for _, op := range mapping.AccessOps {
-			if op.Operation.AccessType != sdkacltypes.AccessType_COMMIT && op.Operation.IdentifierTemplate == "*" {
-				return DefaultGasMultiplierDenominator
-			}
+		msgName := ""
+		msgInfo, err := acltypes.NewExecuteMessageInfo(wasmExecuteMsg.Msg)
+		if err == nil {
+			msgName = msgInfo.MessageName
+		}
+		if messageContainsNoWildcardDependencies(
+			ctx,
+			aclKeeper,
+			wasmExecuteMsg.Contract,
+			msgName,
+			sdkacltypes.WasmMessageSubtype_EXECUTE,
+			make(map[string]struct{}),
+		) {
+			return WasmCorrectDependencyDiscountDenominator
 		}
-		return WasmCorrectDependencyDiscountDenominator
 	}
 	return DefaultGasMultiplierDenominator
 }
+
+// TODO: add tracing to measure latency
+func messageContainsNoWildcardDependencies(
+	ctx sdk.Context,
+	aclKeeper aclkeeper.Keeper,
+	contractAddrStr string,
+	messageName string,
+	messageSubtype sdkacltypes.WasmMessageSubtype,
+	cycleDetector map[string]struct{},
+) bool {
+	cycleIdentifier := fmt.Sprintf("%s:%s", contractAddrStr, messageName)
+	if _, ok := cycleDetector[cycleIdentifier]; ok {
+		// no need to raise an error here. Simply returning false since a contract-message
+		// with cycle will not be parallelized anyway
+		return false
+	}
+	cycleDetector[cycleIdentifier] = struct{}{}
+	addr, err := sdk.AccAddressFromBech32(contractAddrStr)
+	if err != nil {
+		return false
+	}
+	mapping, err := aclKeeper.GetRawWasmDependencyMapping(ctx, addr)
```

Comment:
why not simply call GetWasmDependencyMapping instead of getting the raw one and having branching logic for contract references? That way it would traverse the references for us using the existing logic and wouldnt require repeating logic here, right?

Reply:
The logic here would return early as soon as it hits a *, but agreed that with the new reference logic the repetition is probably too much to be worthwhile the gain. Will update